File: backend/app/models/job.py
from datetime import datetime
from typing import Optional, List
from beanie import Document
from pydantic import Field, BaseModel
from bson import ObjectId
from enum import Enum


class JobType(str, Enum):
    FULL_TIME = "full_time"
    PART_TIME = "part_time"
    CONTRACT = "contract"
    INTERNSHIP = "internship"


class ExperienceLevel(str, Enum):
    ENTRY = "entry"
    MID = "mid"
    SENIOR = "senior"
    LEAD = "lead"
    EXECUTIVE = "executive"


class JobStatus(str, Enum):
    DRAFT = "draft"
    ACTIVE = "active"
    PAUSED = "paused"
    CLOSED = "closed"


class JobLocation(BaseModel):
    city: Optional[str] = None
    state: Optional[str] = None
    country: str = "USA"
    is_remote: bool = False
    remote_type: Optional[str] = None  # "remote", "hybrid", "on_site"


class Job(Document):
    employer_id: ObjectId
    
    # Job Details
    title: str
    description: str
    requirements: List[str] = []
    responsibilities: List[str] = []
    
    # Job Specs
    job_type: JobType
    experience_level: ExperienceLevel
    skills_required: List[str] = []
    education_required: Optional[str] = None
    
    # Compensation
    salary_min: Optional[int] = None
    salary_max: Optional[int] = None
    salary_currency: str = "USD"
    
    # Location
    location: JobLocation
    
    # Employment Details
    industry: Optional[str] = None
    department: Optional[str] = None
    
    # Application Settings
    application_deadline: Optional[datetime] = None
    spots_available: Optional[int] = None
    
    # Status
    status: JobStatus = JobStatus.ACTIVE
    
    # AI Embeddings
    job_embedding: Optional[List[float]] = None
    
    # Metadata
    views_count: int = 0
    applications_count: int = 0
    
    posted_at: datetime = Field(default_factory=datetime.utcnow)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    class Settings:
        name = "jobs"
        indexes = [
            "employer_id",
            "status",
            "job_type",
            "experience_level",
            "skills_required",
            "location.city",
            "posted_at",
        ]

File: frontend/types/job.ts
/**
 * Job-related TypeScript types for the Job Portal
 */

export enum WorkType {
  REMOTE = 'remote',
  HYBRID = 'hybrid',
  ONSITE = 'onsite',
}

export enum JobType {
  FULL_TIME = 'full_time',
  PART_TIME = 'part_time',
  CONTRACT = 'contract',
  INTERNSHIP = 'internship',
}

export enum ExperienceLevel {
  ENTRY = 'entry',
  MID = 'mid',
  SENIOR = 'senior',
  LEAD = 'lead',
}

export enum CompanySize {
  STARTUP = 'startup',
  SMALL = 'small',
  MEDIUM = 'medium',
  LARGE = 'large',
  ENTERPRISE = 'enterprise',
}

export enum JobStatus {
  ACTIVE = 'active',
  ARCHIVED = 'archived',
  DRAFT = 'draft',
}

export interface Job {
  id: string;
  title: string;
  description: string;
  skills: string[];
  employer_id?: string;
  status: JobStatus;
  
  // Location
  location?: string;
  city?: string;
  state?: string;
  country: string;
  
  // Work arrangement
  work_type: WorkType;
  job_type: JobType;
  experience_level: ExperienceLevel;
  easy_apply: boolean;
  
  // Salary
  salary_min?: number;
  salary_max?: number;
  salary_currency: string;
  
  // Company
  company_name?: string;
  company_rating?: number;
  company_size?: CompanySize;
  industry?: string;
  
  // Timestamps
  posted_at: string;
  created_at: string;
  updated_at: string;
  archived_at?: string;
}

export interface SearchFilters {
  // Keyword search
  q?: string;
  location?: string;
  radius?: number;
  
  // Toggle filters
  easy_apply?: boolean;
  remote_only?: boolean;
  
  // Salary
  salary_min?: number;
  salary_max?: number;
  hide_without_salary?: boolean;
  
  // Date posted
  posted_within?: '24h' | '7d' | '30d' | 'any';
  
  // Company
  min_rating?: number;
  company_sizes?: string[];
  companies?: string[];
  industries?: string[];
  
  // Job details
  work_types?: string[];
  job_types?: string[];
  experience_levels?: string[];
  
  // Location
  cities?: string[];
  states?: string[];
  
  // Skills
  skills?: string[];
  
  // Pagination
  page?: number;
  page_size?: number;
  
  // Sort
  sort_by?: 'relevance' | 'newest' | 'salary';
}

export interface PaginationMetadata {
  page: number;
  page_size: number;
  total_results: number;
  total_pages: number;
  has_more: boolean;
}

export interface FilterCount {
  value: string;
  label: string;
  count: number;
}

export interface FilterOptions {
  work_types: FilterCount[];
  job_types: FilterCount[];
  experience_levels: FilterCount[];
  company_sizes: FilterCount[];
  cities: FilterCount[];
  states: FilterCount[];
  salary_ranges: {
    min: number;
    max: number | null;
    label: string;
  }[];
}

export interface SearchResponse {
  jobs: Job[];
  pagination: PaginationMetadata;
  filters_applied: Record<string, any>;
}

